caikit/core/task.py

```diff
@@ -112,6 +104,47 @@ def get_output_type(cls) -> Type[DataBase]:
         """
         raise NotImplementedError("This is implemented by the @task decorator!")
 
+    @classmethod
+    def _raise_on_wrong_output_type(cls, output_type, module):
+        if cls._subclass_check(output_type, cls.get_output_type()):
+            # Basic case, same type or subclass of it
+            return
+
+        if typing.get_origin(output_type) == Union:
+            for union_type in typing.get_args(output_type):
+                if cls._subclass_check(union_type, cls.get_output_type()):
+                    # Something in the union has an acceptable type
+                    return
+
+        # Iterable output checks:
+        # `output_type` here can be any iterable, so long as it is type annotated with the type
+        # that is iterated over. So we first only check that it is iterable at all, and then look
+        # at the annotated type's args.
+        if isiterable(output_type) and isiterable(cls.get_output_type()):
+            # in this case we validated up-front that cls.output_type is Iterable[T]
+            streaming_type = typing.get_args(cls.get_output_type())[0]
```

Comment:
This is error prone if a user uses `List` (with no args) as the type hint

```suggestion
        # at the annotated type's args.
        task_output_type = cls.get_output_type()
        task_output_type_args = get_args(task_output_type)
        if isiterable(output_type) and isiterable(task_output_type) and task_output_type_args:
            # in this case we validated up-front that cls.output_type is Iterable[T]
            streaming_type = task_output_type_args[0]
```

Reply:
Ah! I misunderstood that comment. Looks reasonable as is then